File: netscout.py
# This Python file uses the following encoding: utf-8

# MODULES AND/OR LIBRARIES
from argparse import ArgumentParser
from sys import exit as sysexit
from modules.logging_config import(
    debug,
    info,
    error,
)
from modules.arp_scanner import run_arp_scan
from modules.tcp_scanner import run_tcp_scan
from modules.udp_scanner import run_udp_scan

##############################

# NetScout

##############################

class NetScout:
    def __init__(
        self,
        target_ip_address,
        target_port_specification,
        is_arp_scan_requested,
        is_tcp_scan_requested,
        is_udp_scan_requested
    ):
        self.max_threads = 100
        self.target_ip_address = target_ip_address
        self.target_port_specification = target_port_specification
        self.is_arp_scan_requested = is_arp_scan_requested
        self.is_tcp_scan_requested = is_tcp_scan_requested
        self.is_udp_scan_requested = is_udp_scan_requested

    ##############################

    # ARP SCAN

    ##############################

    def perform_arp_scan(self):
        run_arp_scan(target_address=self.target_ip_address)

    ##############################

    # TCP SCAN

    ##############################

    def perform_tcp_scan(self):
        run_tcp_scan(
            target_ip=self.target_ip_address,
            ports=self.target_port_specification,
            max_threads=self.max_threads
        )

    ##############################

    # UDP SCAN

    ##############################

    def perform_udp_scan(self, timeout=3):
        run_udp_scan(
            target_ip=self.target_ip_address,
            ports=self.target_port_specification,
            max_threads=self.max_threads
        )

    ##############################

    # START

    ##############################

    def start(self):
        if self.is_arp_scan_requested:
            self.perform_arp_scan()
        if self.is_tcp_scan_requested:
            self.perform_tcp_scan()
        if self.is_udp_scan_requested:
            self.perform_udp_scan()


##############################

# HELPER FUNCTIONS

##############################

def parse_port_specification(parser, args):
    if args.port:
        return [args.port]
    elif args.port_range:
        try:
            start, end = map(int, args.port_range.split('-'))
            if start > end:
                parser.error("--port-range must be in the format 'start-end' where start <= end")
                sysexit()
            return list(range(start, end + 1))
        except ValueError:
            parser.error("--port-range must be in the format 'start-end'")
    else:
        return list(range(0, 1001))

##############################

# MAIN

##############################

def main():
    parser = ArgumentParser(description="NetScout: Port & ARP Scanner")
    parser.add_argument("ip_address", help="Target IP address (e.g., 10.10.10.10 or 192.168.1.0/24).")
    port_group = parser.add_mutually_exclusive_group(required=False)
    port_group.add_argument("--port", type=int, help="Target port (e.g., 22).")
    port_group.add_argument("--port-range", help="Target port range (e.g., 0-1000).")
    scan_group = parser.add_mutually_exclusive_group(required=False)
    scan_group.add_argument('-st', action="store_true", help="Scan for open TCP ports on the target.")
    scan_group.add_argument('-sa', action="store_true", help="Perform an ARP scan to discover hosts on the local network.")
    scan_group.add_argument('-su', action="store_true", help="Scan for open UDP ports on the target.")
    args = parser.parse_args()
    target_port_specification = parse_port_specification(parser, args)
    netscout = NetScout(
        target_ip_address=args.ip_address,
        target_port_specification=target_port_specification,
        is_arp_scan_requested=args.sa,
        is_tcp_scan_requested=args.st,
        is_udp_scan_requested=args.su,
    )
    netscout.start()

if __name__ == "__main__":
    main()

File: netscout.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"netscout/modules"
)

// ##############################
//
// # HELPER FUNCTIONS
//
// ##############################

func parsePortRange(rangeStr string) ([]int, error) {
	parts := strings.Split(rangeStr, "-")
	if len(parts) != 2 {
		return nil, fmt.Errorf("invalid port range format. Use <start>-<end>")
	}
	start, err1 := strconv.Atoi(parts[0])
	end, err2 := strconv.Atoi(parts[1])
	if err1 != nil || err2 != nil {
		return nil, fmt.Errorf("invalid number in port range")
	}
	if start < 0 || end > 65535 || start > end {
		return nil, fmt.Errorf("port range must be within 0-65535 and start <= end")
	}

	ports := make([]int, 0, end-start+1)
	for i := start; i <= end; i++ {
		ports = append(ports, i)
	}
	return ports, nil
}

// ##############################
//
// # MAIN
//
// ##############################

func main() {
	ipAddress := flag.String("ip", "", "Target IP address.")
	singlePort := flag.Int("p", -1, "Target port (e.g., -p=22).")
	portRange := flag.String("pr", "", "Target port range (e.g., -pr=0-1000).")
	arpScan := flag.Bool("sa", false, "Perform an ARP scan to discover hosts on the local network.")
	tcpScan := flag.Bool("st", false, "Scan for open TCP ports on the target.")
	udpScan := flag.Bool("su", false, "Scan for open UDP ports on the target.")
	flag.Usage = func() {
		fmt.Println("Usage: netscout [flags]")
		flag.PrintDefaults()
		fmt.Println("\nExamples:")
		fmt.Println("  netscout -ip=10.10.10.10 -st")
		fmt.Println("  netscout -ip=10.10.10.10/24 -sa")
		fmt.Println("  netscout -ip=10.10.10.10 -st -p=3306")
		fmt.Println("  netscout -ip=10.10.10.10 -st -pr=0-4000")
		fmt.Println("  netscout -ip=10.10.10.10 -su -p=9999")
		fmt.Println("  netscout -ip=10.10.10.10 -su -pr=0-4000")
	}
	flag.Parse()

	if *ipAddress == "" {
		fmt.Println("Error: -ip is required.")
		flag.Usage()
		os.Exit(1)
	}

	// Enforce mutual exclusivity of scan modes
	selectedCountScanGroup := 0
	if *arpScan {
		selectedCountScanGroup++
	}
	if *tcpScan {
		selectedCountScanGroup++
	}
	if *udpScan {
		selectedCountScanGroup++
	}
	if selectedCountScanGroup == 0 {
		fmt.Println("Error: At least one of -sa, -st, or -su must be specified.")
		flag.Usage()
		os.Exit(1)
	}
	if selectedCountScanGroup > 1 {
		fmt.Println("Error: Only one of -sa, -st, or -su can be specified at a time.")
		flag.Usage()
		os.Exit(1)
	}

	// Enforce mutual exclusivity of port specs if scanning ports
	selectedCountPortSpec := 0
	if *singlePort != -1 {
		selectedCountPortSpec++
	}
	if *portRange != "" {
		selectedCountPortSpec++
	}
	if selectedCountPortSpec > 1 {
		fmt.Println("Error: Only one of -p or -pr can be specified.")
		flag.Usage()
		os.Exit(1)
	}

	// Prepare ports list
	var ports []int
	if *singlePort != -1 {
		if *singlePort < 0 || *singlePort > 65535 {
			fmt.Println("Error: Port must be between 0-65535.")
			os.Exit(1)
		}
		ports = []int{*singlePort}
	} else if *portRange != "" {
		parsedPorts, err := parsePortRange(*portRange)
		if err != nil {
			fmt.Println("Error:", err)
			os.Exit(1)
		}
		ports = parsedPorts
	} else {
		// Default ports
		for i := 0; i <= 1000; i++ {
			ports = append(ports, i)
		}
	}

	// Max threads constant (same as Python version)
	const maxThreads = 100

	// Run requested modules
	if *arpScan {
		modules.RunArpScan(*ipAddress)
	}

	if *tcpScan {
		// Load IANA service name-port mapping before any TCP scan
		csvPath := "data/service-names-port-numbers.csv"
		if err := modules.LoadIanaServices(csvPath); err != nil {
			fmt.Printf("[!] Could not load service names CSV (%s): %v\n", csvPath, err)
		}

		modules.RunTcpScan(*ipAddress, ports, maxThreads)
	}

	if *udpScan {
		modules.CheckRootPrivileges()
		timeout := 5 * time.Second
		quiet := 3 * time.Second
		maxWait := 10 * time.Second
		modules.RunUDPScan(*ipAddress, ports, maxThreads, timeout, quiet, maxWait)
	}
}